python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/_tracer.py

```diff
@@ -62,30 +67,73 @@ def audit_timing(wrapped: Any, _: Any, args: Any, kwargs: Any) -> Any:
         print(f"{wrapped.__name__}: {latency_ms:.2f}ms")
 
 
-class _Run(NamedTuple):
-    span: trace_api.Span
-    context: context_api.Context
+K = TypeVar("K", bound=Hashable)
+V = TypeVar("V")
+
+
+class _DictWithLock(ObjectProxy, Generic[K, V]):  # type: ignore
+    """
+    A wrapped dictionary with lock
+    """
+
+    def __init__(self, wrapped: Optional[Dict[str, V]] = None) -> None:
+        super().__init__(wrapped or {})
+        self._self_lock = RLock()
+
+    def get(self, key: K) -> Optional[V]:
+        with self._self_lock:
+            return cast(Optional[V], self.__wrapped__.get(key))
+
+    def pop(self, key: K, *args: Any) -> Optional[V]:
+        with self._self_lock:
+            return cast(Optional[V], self.__wrapped__.pop(key, *args))
+
+    def __getitem__(self, key: K) -> V:
+        with self._self_lock:
+            return cast(V, super().__getitem__(key))
+
+    def __setitem__(self, key: K, value: V) -> None:
+        with self._self_lock:
+            super().__setitem__(key, value)
+
+    def __delitem__(self, key: K) -> None:
+        with self._self_lock:
+            super().__delitem__(key)
 
 
 class OpenInferenceTracer(BaseTracer):
-    __slots__ = ("_tracer", "_runs")
+    __slots__ = ("_tracer", "_spans", "_initialized")
+    _singleton: Optional["OpenInferenceTracer"] = None
```

Comment:
probably add a doctoring explaining the reason this singleton pointer is needed

Reply:
I changed my mind about making it a Singleton (capitalized) class because that's more complicated than necessary. So i changed it to just a `singleton` (lower case) instance in the `instrument()` step instead.